In [2]:
#importing packages
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import seaborn as sns
import statsmodels.formula.api as smf
import statsmodels.api as sm

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [3]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.metrics import confusion_matrix
from sklearn import preprocessing
from sklearn.model_selection import train_test_split

In [4]:
data=pd.read_csv("bank-full.csv", sep=';')
data.head()

age           job  marital  education  ... pdays  previous poutcome   y
0   58    management  married   tertiary  ...    -1         0  unknown  no
1   44    technician   single  secondary  ...    -1         0  unknown  no
2   33  entrepreneur  married  secondary  ...    -1         0  unknown  no
3   47   blue-collar  married    unknown  ...    -1         0  unknown  no
4   33       unknown   single    unknown  ...    -1         0  unknown  no

[5 rows x 17 columns]

In [5]:
data.shape

(45211, 17)

In [6]:
data=data.dropna()
data.shape

(45211, 17)

In [8]:
data.isnull().sum()

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
y            0
dtype: int64

In [9]:
x=data.iloc[:,:16]
y=data.iloc[:, 16]

In [10]:
x=pd.get_dummies(x)
x

age  balance  day  ...  poutcome_other  poutcome_success  poutcome_unknown
0       58     2143    5  ...               0                 0                 1
1       44       29    5  ...               0                 0                 1
2       33        2    5  ...               0                 0                 1
3       47     1506    5  ...               0                 0                 1
4       33        1    5  ...               0                 0                 1
...    ...      ...  ...  ...             ...               ...               ...
45206   51      825   17  ...               0                 0                 1
45207   71     1729   17  ...               0                 0                 1
45208   72     5715   17  ...               0                 1                 0
45209   57      668   17  ...               0                 0                 1
45210   37     2971   17  ...               1                 0                 0

[45211 rows x 51 columns]

In [11]:
classifier=LogisticRegression()
classifier.fit(x,y)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [12]:
y_pred=classifier.predict(x)

In [14]:
y_pred_df= pd.DataFrame({'actual': y,'predicted_prob': classifier.predict(x)})
y_pred_df

actual predicted_prob
0         no             no
1         no             no
2         no             no
3         no             no
4         no             no
...      ...            ...
45206    yes            yes
45207    yes             no
45208    yes            yes
45209     no             no
45210     no             no

[45211 rows x 2 columns]

In [15]:
from sklearn.metrics import confusion_matrix
confusion_matrix=confusion_matrix(y,y_pred)
print(confusion_matrix)

[[39145   777]
 [ 4114  1175]]


In [17]:
((39145+1175)/(39145+777+4114+1175))*100

89.18183627878172

In [18]:
from sklearn.metrics import classification_report
print(classification_report(y,y_pred))

              precision    recall  f1-score   support

          no       0.90      0.98      0.94     39922
         yes       0.60      0.22      0.32      5289

    accuracy                           0.89     45211
   macro avg       0.75      0.60      0.63     45211
weighted avg       0.87      0.89      0.87     45211



# hence as we can see our model is showing accuracy of **89%** in it